src/mscorlib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -128,134 +112,239 @@ public static bool IsFlowSuppressed()
             return executionContext != null && executionContext.m_isFlowSuppressed;
         }
 
+        internal bool HasChangeNotifications => m_localChangeNotifications != null;
+
         public static void Run(ExecutionContext executionContext, ContextCallback callback, Object state)
         {
+            // Note: ExecutionContext.Run is an extremely hot function and used by every await, ThreadPool execution, etc.
             if (executionContext == null)
-                throw new InvalidOperationException(SR.InvalidOperation_NullContext);
+            {
+                ThrowNullContext();
```

Comment:
There are several places in corelib where we call this knowing it's not null. Is the JIT able to remove the check in such cases? Or if not, would we gain anything measurable by having an internal version of this without the check?

Reply:
Can split into two one with check one without; e.g. Run -> RunInternal; Run will then inline and remove the check; however also can call RunInternal directly. Will do that.

Reply:
Could add null check to Default test and ThreadPool `QueueUserWorkItemCallbackDefaultContext` could call
```csharp
// null context on RunInternal is Default context
ExecutionContext.RunInternal(context: null, ccb, this);
```
instead of
```chsarp
ExecutionContext.RunInternal(ExecutionContext.Default, ccb, this);
```
And save static helper lookup?

Reply:
Changed that also